File: backend/src/main/java/blok2/model/stats/LocationOverviewStat.java
package blok2.model.stats;

import org.springframework.boot.configurationprocessor.json.JSONObject;

import javax.persistence.*;
import java.util.Map;
import java.util.stream.Collectors;

@Entity
public class LocationOverviewStat {

    @Id
    private int locationId;

    private String locationName;

    private long reservationsTotal;

    @ElementCollection
    @MapKeyColumn(name = "HOI")
    @Column(name = "count")
    private Map<String, Long> reservationsTotalPerHOI;

    @ElementCollection
    @MapKeyColumn(name = "date")
    @Column(name = "count")
    private Map<String, Long> reservationsPerDay;

    @ElementCollection
    @MapKeyColumn(name = "date")
    @Column(name = "map")
    private Map<String, String> reservationsPerDayPerHOI;

    public LocationOverviewStat(int locationId, String locationName, long reservationsTotal, Map<String, Long> reservationsTotalPerHOI, Map<String, Long> reservationsPerDay, Map<String, Map<String, Long>> reservationsPerDayPerHOI) {
        this.locationId = locationId;
        this.locationName = locationName;
        this.reservationsTotal = reservationsTotal;
        this.reservationsTotalPerHOI = reservationsTotalPerHOI;
        this.reservationsPerDay = reservationsPerDay;
        this.reservationsPerDayPerHOI = reservationsPerDayPerHOI.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, e -> new JSONObject(e.getValue()).toString()));
    }

    public LocationOverviewStat() {
    }

    public int getLocationId() {
        return locationId;
    }

    public String getLocationName() {
        return locationName;
    }

    public long getReservationsTotal() {
        return reservationsTotal;
    }

    public Map<String, Long> getReservationsTotalPerHOI() {
        return reservationsTotalPerHOI;
    }

    public Map<String, Long> getReservationsPerDay() {
        return reservationsPerDay;
    }

    public Map<String, String> getReservationsPerDayPerHOI() {
        return reservationsPerDayPerHOI;
    }
}

File: frontend/src/app/shared/model/LocationOverviewStat.ts
export interface LocationOverviewStat {
    locationId: number;
    locationName: string;
    reservationsTotal: number;
    reservationsTotalPerHOI: Map<string, number>;
    reservationsPerDay: Map<string, number>;
    reservationsPerDayPerHOI: Map<string, Map<string, number>>;
}

export class LocationOverviewStatConstructor {
    static new(): LocationOverviewStat {
        return {
            locationId: -1,
            locationName: '',
            reservationsTotal: 0,
            reservationsTotalPerHOI: new Map<string, number>(),
            reservationsPerDay: new Map<string, number>(),
            reservationsPerDayPerHOI: new Map<string, Map<string, number>>()
        };
    }

    static newFromObj(obj: LocationOverviewStat): LocationOverviewStat {
        if (obj === null) {
            return null;
        }

        const reservationsTotalPerHOI = new Map(Object.entries(obj.reservationsTotalPerHOI));
        const reservationsPerDay = new Map(Object.entries(obj.reservationsPerDay));
        const reservationsPerDayPerHOI = new Map([...Object.entries(obj.reservationsPerDayPerHOI)].sort((a, b) => {
            return new Date(a[0]).getTime() - new Date(b[0]).getTime();
        }).map(e => {
            return [e[0], JSON.parse(e[1])];
        }));

        return {
            locationId: obj.locationId,
            locationName: obj.locationName,
            reservationsTotal: obj.reservationsTotal,
            reservationsTotalPerHOI: reservationsTotalPerHOI,
            reservationsPerDay: reservationsPerDay,
            reservationsPerDayPerHOI: reservationsPerDayPerHOI
        };
    }
}
